### PR TITLE
fix: use single wasm instance

### PR DESCRIPTION
**Pull request by ronag**
Fixes: https://github.com/nodejs/undici/issues/725

### CODE REVIEW

**Comment by ronag** [comment]
I'll quickly publish another alpha once this has been merged.

File: lib/client.js [hunk 1]
```diff
@@ -402,57 +402,57 @@ class HTTPParserError extends Error {
   }
 }
 
+let currentParser = null
+const llhttp = new WebAssembly.Instance(mod, {
```
Comment (dnlup) [inline lib/client.js:406]: I think here we can specify memory limits, if we need it in the future

**Review by dnlup — APPROVED** [review]

File: lib/client.js (hunk 1 above)
Comment (ronag) [reply lib/client.js:406]: Do we know what the default limits are?

Comment (dnlup) [reply lib/client.js:406]: I couldn't find it yet. Maybe a manual inspection could tell. But it should be easy to pass a custom memory instance, we just need to add a `memory` key to `env` if I am not mistaken:

* https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Memory/Memory#parameters
* https://ariya.io/2019/05/basics-of-memory-access-in-webassembly

Btw, I am not saying we should do it now.

**Comment by mcollina** [comment]
I don't think this is correct. We have one parser per req/res in node core as well.

**Review by mcollina — APPROVED** [review]
lgtm anyway.

It would be better to have a regression test

**Comment by dnlup** [comment]
> I don't think this is correct. We have one parser per req/res in node core as well.

I understand. Just the wasm instance would be shared (and the memory instantiated just once). The `js` parser and the `llhttp` would be one for each req/res, if I am not mistaken.

File: lib/client.js [hunk 2]
```diff
@@ -831,6 +844,16 @@ class Parser {
   }
 
   destroy () {
+    if (this.ptr) {
+      llhttp.llhttp_free(this.ptr)
```
Comment (dnlup) [inline lib/client.js:848]: Maybe  we were not freeing memory before?

Comment (ronag) [reply lib/client.js:848]: Wasn't needed since the wasm was thrown away.

**Comment by mcollina** [comment]
Good, then let's land this!

**Pull request merged**